HPCC-29191 Enable stack trace when no credentials provided to Dali

RussWhitehead opened the pull request:
Log stack trace when Dali file access requests do not provide credentials. Do this no more than once a minute, to ensure logfile not flooded

<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [x] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

## Testing:
<!-- Please describe how this change has been tested.-->

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

richardkchapman committed:
Some implementation notes on Roxie for the benefit of future maintainers

Signed-off-by: Richard Chapman <rchapman@hpccsystems.com>

ghalliday committed:
HPCC-29029 Add guidelines for creating pull requests

Signed-off-by: Gavin Halliday <gavin.halliday@lexisnexis.com>

ghalliday committed:
Split off 9.0.0

Signed-off-by: Gavin Halliday <gavin.halliday@lexisnexis.com>

ghalliday committed:
Merge remote-tracking branch 'origin/candidate-9.0.x'

Signed-off-by: Gavin Halliday <gavin.halliday@lexisnexis.com>

# Conflicts:
#	helm/hpcc/Chart.yaml
#	helm/hpcc/templates/_helpers.tpl
#	helm/hpcc/templates/dafilesrv.yaml
#	helm/hpcc/templates/dali.yaml
#	helm/hpcc/templates/dfuserver.yaml
#	helm/hpcc/templates/eclagent.yaml
#	helm/hpcc/templates/eclccserver.yaml
#	helm/hpcc/templates/eclscheduler.yaml
#	helm/hpcc/templates/esp.yaml
#	helm/hpcc/templates/localroxie.yaml
#	helm/hpcc/templates/roxie.yaml
#	helm/hpcc/templates/sasha.yaml
#	helm/hpcc/templates/thor.yaml
#	version.cmake

ghalliday committed:
Merge remote-tracking branch 'origin/candidate-9.0.x'

Signed-off-by: Gavin Halliday <gavin.halliday@lexisnexis.com>

RussWhitehead committed:
HPCC-29191 Enable stack trace when no credentials provided to Dali

Log stack trace when Dali file access requests do not provide credentials.
Do this no more than once a minute, to ensure logfile not flooded

Signed-off-by: Russ Whitehead <russwhitehead@yahoo.com>

jakesmith reviewed (changes requested):
@RussWhitehead - see comments.

ghalliday committed:
Merge pull request #16949 from richardkchapman/roxie-doc

Some implementation notes on Roxie for the benefit of future maintainers

Reviewed-by: Mark Kelly mark.kelly@lexisnexisrisk.com
Reviewed-by: Gavin Halliday <ghalliday@hpccsystems.com>
Merged-by: Gavin Halliday <ghalliday@hpccsystems.com>

RussWhitehead commented on dali/base/CMakeLists.txt:
```diff
@@ -72,7 +72,7 @@ include_directories (
     )
 
-ADD_DEFINITIONS( -D_USRDLL -DDALI_EXPORTS )
+ADD_DEFINITIONS( -D_USRDLL -DDALI_EXPORTS -DNULL_DALIUSER_STACKTRACE)
```
@jakesmith The NULL_DALIUSER_STACKTRACE tracing only exists in dasess.cpp and dadfs.cpp, and this PR takes care of them all

RussWhitehead commented on dali/base/dadfs.cpp:
```diff
@@ -1341,9 +1347,15 @@ static void setUserDescriptor(Linked<IUserDescriptor> &udesc,IUserDescriptor *us
+            CriticalBlock block(nullUserLogCS);
+            time_t timeNow;
+            time(&timeNow);
+            if (difftime(timeNow, lastNullUserLogEntry) >= nullUserLogMinutes)
```
@jakesmith I am not familiar with "func", but since this is temporary tracing code I hope we can do without it for now

jakesmith commented on dali/base/dasess.cpp:
```diff
@@ -151,6 +151,26 @@ static CriticalSection sessionCrit;
+{
+    const double nullUserLogMinutes = 60;	//log NULL user requests no more than once this often
+    static time_t lastNullUserLogEntry = (time_t)NULL;
+    static CriticalSection nullUserLogCS;
```
local static objects come with a hidden cost, c++ standard guarantees that they are initialized in a thread-safe way, meaning there's an implicit lock involved here.

To avoid the overhead, it is better to move these statics out of the function into module scope.

jakesmith commented on dali/base/dasess.cpp:
```diff
@@ -151,6 +151,26 @@ static CriticalSection sessionCrit;
+static void logNullUser(const char * func, int line = __LINE__)
+{
+    const double nullUserLogMinutes = 60;	//log NULL user requests no more than once this often
+    static time_t lastNullUserLogEntry = (time_t)NULL;
```
time_t is a scalar, don't think nullptr should be used here, 0 would be more correct.

jakesmith reviewed (changes requested):
@RussWhitehead - please see comments.

ghalliday committed:
Merge remote-tracking branch 'origin/candidate-8.12.x'

Signed-off-by: Gavin Halliday <gavin.halliday@lexisnexis.com>

# Conflicts:
#	helm/hpcc/Chart.yaml
#	helm/hpcc/templates/_helpers.tpl
#	helm/hpcc/templates/dafilesrv.yaml
#	helm/hpcc/templates/dali.yaml
#	helm/hpcc/templates/dfuserver.yaml
#	helm/hpcc/templates/eclagent.yaml
#	helm/hpcc/templates/eclccserver.yaml
#	helm/hpcc/templates/eclscheduler.yaml
#	helm/hpcc/templates/esp.yaml
#	helm/hpcc/templates/localroxie.yaml
#	helm/hpcc/templates/roxie.yaml
#	helm/hpcc/templates/sasha.yaml
#	helm/hpcc/templates/thor.yaml
#	version.cmake

ghalliday committed:
Merge remote-tracking branch 'origin/candidate-9.0.x'

Signed-off-by: Gavin Halliday <gavin.halliday@lexisnexis.com>

# Conflicts:
#	helm/hpcc/Chart.yaml
#	helm/hpcc/templates/_helpers.tpl
#	helm/hpcc/templates/dafilesrv.yaml
#	helm/hpcc/templates/dali.yaml
#	helm/hpcc/templates/dfuserver.yaml
#	helm/hpcc/templates/eclagent.yaml
#	helm/hpcc/templates/eclccserver.yaml
#	helm/hpcc/templates/eclscheduler.yaml
#	helm/hpcc/templates/esp.yaml
#	helm/hpcc/templates/localroxie.yaml
#	helm/hpcc/templates/roxie.yaml
#	helm/hpcc/templates/sasha.yaml
#	helm/hpcc/templates/thor.yaml
#	version.cmake

shamser committed:
HPCC-28837 Extend FileSpray ESP to support destinationNumParts

Signed-off-by: Shamser Ahmed <shamser.ahmed@lexisnexis.com>

jakesmith committed:
HPCC-29282 Show Dali size and load progress

Signed-off-by: Jake Smith <jake.smith@lexisnexisrisk.com>

ghalliday committed:
HPCC-28874 Remove strncpy() when processing x'...' with odd number of digits

Signed-off-by: Gavin Halliday <gavin.halliday@lexisnexis.com>

ghalliday committed:
HPCC-29312 Allow more granular control of eclcc advanced tracing

* Add -ftraceAll=true|false to enable all detailed tracing
* Add -ftraceX=true|false to control tracing of feature X

Signed-off-by: Gavin Halliday <gavin.halliday@lexisnexis.com>

richardkchapman committed:
HPCC-28652 Enable NEW_IBYTI in bare-metal

Signed-off-by: Richard Chapman <rchapman@hpccsystems.com>

richardkchapman committed:
HPCC-28068 Demand load bloom filters in indexes

Signed-off-by: Richard Chapman <rchapman@hpccsystems.com>

richardkchapman committed:
HPCC-27526 Avoid reading any non leaf nodes when reading an index with no filter

Prep work refactoring that would be required if we were to try to make the loading
of the root node be on-demand. This change is a worthwhile code cleanup anyway, but
I'm not convinced it makes sense to complete the work since unfiltered index reading
is a very rare operation compared to filtered reads, which would incur an overhead
on every seek.

Signed-off-by: Richard Chapman <rchapman@hpccsystems.com>

ghalliday committed:
Merge pull request #17215 from richardkchapman/root-on-demand

HPCC-27526 Avoid reading any non leaf nodes when reading an index with no filter

Reviewed-by: Gavin Halliday <ghalliday@hpccsystems.com>
Merged-by: Gavin Halliday <ghalliday@hpccsystems.com>

richardkchapman committed:
HPCC-29347 Out-of-date comments in CKeyIndex::getRootNode

Signed-off-by: Richard Chapman <rchapman@hpccsystems.com>

ghalliday committed:
Merge remote-tracking branch 'origin/candidate-9.0.x'

Signed-off-by: Gavin Halliday <gavin.halliday@lexisnexis.com>

ghalliday committed:
Merge pull request #17216 from richardkchapman/update-comments

HPCC-29347 Out-of-date comments in CKeyIndex::getRootNode

Reviewed-by: Gavin Halliday <ghalliday@hpccsystems.com>
Merged-by: Gavin Halliday <ghalliday@hpccsystems.com>

ghalliday committed:
Merge pull request #17196 from ghalliday/issue28874

HPCC-28874 Remove strncpy() when processing x'...' with odd number of digits

Reviewed-By: Shamser Ahmed <shamser.ahmed@lexisnexis.com>
Merged-by: Gavin Halliday <ghalliday@hpccsystems.com>

ghalliday committed:
Merge pull request #17018 from ghalliday/issue29029

HPCC-29029 Add guidelines for creating pull requests

Reviewed-By: Chris Lo
Reviewed-by: Gordon Smith <GordonJSmith@gmail.com>
Reviewed-by: Jake Smith <jake.smith@lexisnexisrisk.com>
Reviewed-By: Richard Chapman <rchapman@hpccsystems.com>
Merged-by: Gavin Halliday <ghalliday@hpccsystems.com>

ghalliday committed:
Merge pull request #17201 from richardkchapman/roxie-multitest

HPCC-28652 Enable NEW_IBYTI in bare-metal

Reviewed-by: Mark Kelly mark.kelly@lexisnexisrisk.com
Reviewed-by: Gavin Halliday <ghalliday@hpccsystems.com>
Merged-by: Gavin Halliday <ghalliday@hpccsystems.com>

richardkchapman committed:
HPCC-29363 Add option to enabled perf flame graph collection per query

Signed-off-by: Richard Chapman <rchapman@hpccsystems.com>

richardkchapman committed:
HPCC-29367 Option to see flame graphs on Thor graphs

Signed-off-by: Richard Chapman <rchapman@hpccsystems.com>

ghalliday committed:
Merge pull request #17227 from richardkchapman/query-flame-2

 HPCC-29363 Add option to enabled perf flame graph collection per query

Reviewed-by: Gavin Halliday <ghalliday@hpccsystems.com>
Merged-by: Gavin Halliday <ghalliday@hpccsystems.com>

ghalliday committed:
Merge remote-tracking branch 'origin/candidate-9.0.x'

Signed-off-by: Gavin Halliday <gavin.halliday@lexisnexis.com>

# Conflicts:
#	helm/hpcc/Chart.yaml
#	helm/hpcc/templates/_helpers.tpl
#	helm/hpcc/templates/dafilesrv.yaml
#	helm/hpcc/templates/dali.yaml
#	helm/hpcc/templates/dfuserver.yaml
#	helm/hpcc/templates/eclagent.yaml
#	helm/hpcc/templates/eclccserver.yaml
#	helm/hpcc/templates/eclscheduler.yaml
#	helm/hpcc/templates/esp.yaml
#	helm/hpcc/templates/localroxie.yaml
#	helm/hpcc/templates/roxie.yaml
#	helm/hpcc/templates/sasha.yaml
#	helm/hpcc/templates/thor.yaml
#	version.cmake

ghalliday committed:
Merge pull request #17198 from ghalliday/issue29312

HPCC-29312 Allow more granular control of eclcc advanced tracing

Reviewed-By: Shamser Ahmed <shamser.ahmed@lexisnexis.com>
Reviewed-By: Richard Chapman <rchapman@hpccsystems.com>
Merged-by: Gavin Halliday <ghalliday@hpccsystems.com>

richardkchapman committed:
HPCC-28079 Revisit roxie cache warming

Some additional notes and thoughts.

Signed-off-by: Richard Chapman <rchapman@hpccsystems.com>

ghalliday committed:
Merge pull request #17191 from jakesmith/HPCC-29282-dali-load-filesize

HPCC-29282 Show Dali size and load progress

Reviewed-by: Gavin Halliday <ghalliday@hpccsystems.com>
Merged-by: Gavin Halliday <ghalliday@hpccsystems.com>

ghalliday committed:
Merge remote-tracking branch 'origin/candidate-9.0.x'

Signed-off-by: Gavin Halliday <gavin.halliday@lexisnexis.com>

ghalliday committed:
Merge pull request #17208 from richardkchapman/bloom-ondemand

HPCC-28068 Demand load bloom filters in indexes

Reviewed-by: Mark Kelly mark.kelly@lexisnexisrisk.com
Reviewed-by: Gavin Halliday <ghalliday@hpccsystems.com>
Merged-by: Gavin Halliday <ghalliday@hpccsystems.com>

ghalliday committed:
Merge pull request #17245 from richardkchapman/roxie-prewarm-doc

HPCC-28079 Revisit roxie cache warming

Reviewed-by: Gavin Halliday <ghalliday@hpccsystems.com>
Merged-by: Gavin Halliday <ghalliday@hpccsystems.com>

ghalliday committed:
Merge pull request #17228 from richardkchapman/thor_flame_2

HPCC-29367 Option to see flame graphs on Thor graphs

Reviewed-by: Jake Smith <jake.smith@lexisnexisrisk.com>
Reviewed-by: Gavin Halliday <ghalliday@hpccsystems.com>
Merged-by: Gavin Halliday <ghalliday@hpccsystems.com>

ghalliday committed:
Merge pull request #17192 from shamser/issue28837

HPCC-28837 Extend FileSpray ESP to support destinationNumParts

Reviewed-by: Jake Smith <jake.smith@lexisnexisrisk.com>
Reviewed-by: Gavin Halliday <ghalliday@hpccsystems.com>
Merged-by: Gavin Halliday <ghalliday@hpccsystems.com>

ghalliday committed:
Merge remote-tracking branch 'origin/candidate-9.0.x'

Signed-off-by: Gavin Halliday <gavin.halliday@lexisnexis.com>

RussWhitehead committed:
Merge remote-tracking branch 'origin/nullUser90x' into nullUser90x

RussWhitehead committed:
Signed-off-by: Russ Whitehead <russwhitehead@yahoo.com>

RussWhitehead commented:
Closing, and replacing with a new pull request
